File: js/tests/lsStream.test.ts
/** LS증권 실시간 스트림 — 로컬 ws 서버로 프로토콜을 재생한다 (Kotlin LsMarketStreamTest 대응, 문서 기반). */
import assert from "node:assert/strict";
import { test } from "node:test";
import { LsMarketStream, parseLsOrderBook, parseLsOrderEvents, parseLsTrade } from "../src/brokers/lsStream.js";
import type { OrderBookTick, OrderEvent, TradeTick } from "../src/models.js";
import { Queue, TODAY, assertBooks, assertEvents, assertTrades, sleep, streamFixture, withServer } from "./wsHarness.js";

const fx = streamFixture("ls");
const TRADE_FRAME = fx.frames[0];
const BOOK_FRAME = fx.orderBook.frames[0];
const [SC0_FRAME, SC1_FRAME, SC3_FRAME] = fx.orderEvents.frames;
const opts = (url: string) => ({ wsUrl: url, token: async () => "TOKEN" });

async function takeN(conn: { received: Queue<string> }, n: number): Promise<{ header: Record<string, string>; body: Record<string, string> }[]> {
  const out: { header: Record<string, string>; body: Record<string, string> }[] = [];
  for (let i = 0; i < n; i++) out.push(JSON.parse(await conn.received.take()));
  return out;
}

test("ls: 종목 하나를 구독하면 KOSPI·KOSDAQ 체결 TR 을 tr_type 3 으로 둘 다 등록하고, 체결 프레임을 요청 표기로 전달한다", () =>
  withServer(async (url, connections) => {
    const ticks = new Queue<TradeTick>();
    const stream = new LsMarketStream(opts(url));
    try {
      stream.subscribeTrades(["KRX:005930"], (t) => ticks.put(t));
      stream.connect();
      const conn = await connections.take();
      const regs = await takeN(conn, 2);
      assert.deepEqual(regs.map((r) => r.body.tr_cd).sort(), ["K3_", "S3_"]);
      assert.ok(regs.every((r) => r.header.token === "TOKEN" && r.header.tr_type === "3" && r.body.tr_key === "005930"));
      conn.send('{"header":{"tr_cd":"S3_","tr_key":"005930","tr_type":"3","rsp_cd":"00000","rsp_msg":"정상처리되었습니다"}}'); // ACK
      conn.send(TRADE_FRAME);
      const tick = await ticks.take();
      assert.equal(tick.symbol, "KRX:005930");
      assert.ok(tick.price.eq(55550));
      assert.ok(tick.quantity.eq(1));
      assert.equal(tick.cumulativeVolume, 10887);
      assert.ok(tick.change!.eq(1050));
      assert.ok(tick.changeRate!.eq("0.0193"));
      assert.ok(tick.askPrice!.eq(55600));
      assert.ok(tick.bidPrice!.eq(55500));
      // KOSDAQ TR 로 같은 레이아웃이 와도 전달된다
      conn.send(TRADE_FRAME.replace('"tr_cd":"S3_"', '"tr_cd":"K3_"'));
      assert.ok((await ticks.take()).price.eq(55550));
    } finally { stream.close(); }
  }));

test("ls: 주문 통보를 구독하면 SC0~SC4 를 tr_type 1·빈 tr_key 로 등록하고, 접수·체결·취소 프레임을 이벤트로 전달한다", () =>
  withServer(async (url, connections) => {
    const events = new Queue<OrderEvent>();
    const stream = new LsMarketStream(opts(url));
    try {
      stream.subscribeOrderEvents((e) => events.put(e));
      stream.connect();
      const conn = await connections.take();
      const regs = await takeN(conn, 5);
      assert.deepEqual(regs.map((r) => r.body.tr_cd).sort(), ["SC0", "SC1", "SC2", "SC3", "SC4"]);
      assert.ok(regs.every((r) => r.header.tr_type === "1" && r.body.tr_key === ""));

      conn.send(SC0_FRAME);
      const accepted = await events.take();
      assert.equal(accepted.type, "ACCEPTED");
      assert.equal(accepted.orderId, "86382");
      assert.equal(accepted.side, "BUY");
      assert.ok(accepted.quantity!.eq(2));
      assert.ok(accepted.price!.eq(60000));
      assert.equal(accepted.symbol, "005930");

      conn.send(SC1_FRAME);
      const filled = await events.take();
      assert.equal(filled.type, "FILLED");
      assert.ok(filled.quantity!.eq(1));
      assert.ok(filled.price!.eq(60000));
      assert.ok(filled.remainingQuantity!.eq(1));

      conn.send(SC3_FRAME);
      const canceled = await events.take();
      assert.equal(canceled.type, "CANCELED");
      assert.equal(canceled.orderId, "88343");
      assert.equal(canceled.originalOrderId, "88342");
      assert.equal(canceled.symbol, "000020");
    } finally { stream.close(); }
  }));

test("ls: 호가 구독은 H1_·HA_ 를 등록하고 호가 프레임을 10단계 호가창으로 전달한다", () =>
  withServer(async (url, connections) => {
    const books = new Queue<OrderBookTick>();
    const stream = new LsMarketStream(opts(url));
    try {
      stream.subscribeOrderBook(["KRX:005930"], (b) => books.put(b));
      stream.connect();
      const conn = await connections.take();
      assert.deepEqual((await takeN(conn, 2)).map((r) => r.body.tr_cd).sort(), ["H1_", "HA_"]);
      conn.send(BOOK_FRAME);
      const book = await books.take();
      assert.equal(book.symbol, "KRX:005930");
      assert.equal(book.asks.length, 10);
      assert.ok(book.asks[0].price.eq(72400));
      assert.ok(book.asks[0].quantity.eq(32616));
      assert.ok(book.bids[0].price.eq(72300));
      assert.ok(book.bids[9].price.eq(71400));
      assert.ok(book.totalAskQuantity!.eq(400000));
      assert.ok(book.totalBidQuantity!.eq(500000));
    } finally { stream.close(); }
  }));

test("ls: 서버가 끊으면 재접속해 모든 등록을 다시 보낸다", () =>
  withServer(async (url, connections) => {
    const stream = new LsMarketStream(opts(url));
    try {
      stream.subscribeTrades(["005930"], () => {});
      stream.subscribeOrderEvents(() => {});
      stream.connect();
      const first = await connections.take();
      await takeN(first, 7);
      first.socket.close(1000, "bye");
      const second = await connections.take(10_000);
      const trs = (await takeN(second, 7)).map((r) => r.body.tr_cd).sort();
      assert.deepEqual(trs, ["K3_", "S3_", "SC0", "SC1", "SC2", "SC3", "SC4"]);
    } finally { stream.close(); }
  }));

test("ls: 연결된 뒤 추가 구독은 즉시, 해제는 tr_type 4·2 로 전송된다", () =>
  withServer(async (url, connections) => {
    const stream = new LsMarketStream(opts(url));
    try {
      stream.subscribeOrderEvents(() => {});
      stream.connect();
      const conn = await connections.take();
      await takeN(conn, 5);
      await sleep(50);
      stream.subscribeTrades(["000660"], () => {});
      const subs = await takeN(conn, 2);
      assert.ok(subs.every((r) => r.body.tr_key === "000660" && r.header.tr_type === "3"));
      stream.unsubscribeTrades(["000660"]);
      const unsubs = await takeN(conn, 2);
      assert.ok(unsubs.every((r) => r.header.tr_type === "4"));
      stream.unsubscribeOrderEvents();
      const unregs = await takeN(conn, 5);
      assert.ok(unregs.every((r) => r.header.tr_type === "2"));
    } finally { stream.close(); }
  }));

test("ls 파서: ACK·오류 프레임은 데이터로 처리하지 않고, 정정·거부·ordxctptncode 우선을 판정한다", () => {
  const ack = { header: { tr_cd: "S3_", tr_key: "005930", tr_type: "3", rsp_cd: "00000", rsp_msg: "정상처리되었습니다" } };
  assert.equal(parseLsTrade(ack, TODAY), null);
  assert.equal(parseLsOrderBook({ ...ack, header: { ...ack.header, tr_cd: "H1_" } }, TODAY), null);
  assert.equal(parseLsOrderEvents({ header: { tr_cd: "SC1", rsp_msg: "x" } }, TODAY).length, 0);

  const sc1 = JSON.parse(SC1_FRAME);
  const modified = parseLsOrderEvents({ header: { tr_cd: "SC2" }, body: { ...sc1.body, ordxctptncode: "12", ordno: "86383", orgordno: "86382", mdfycnfqty: "1", mdfycnfprc: "70000" } }, TODAY)[0];
  assert.equal(modified.type, "MODIFIED");
  assert.ok(modified.price!.eq(70000));
  assert.equal(modified.originalOrderId, "86382");
  const rejected = parseLsOrderEvents({ header: { tr_cd: "SC4" }, body: { ...sc1.body, ordxctptncode: "", rjtqty: "2", msgcode: "1234" } }, TODAY)[0];
  assert.equal(rejected.type, "REJECTED");
  assert.equal(rejected.reason, "1234");
  assert.ok(rejected.quantity!.eq(2));
  // ordxctptncode 가 tr_cd 보다 우선한다
  assert.equal(parseLsOrderEvents({ header: { tr_cd: "SC1" }, body: { ...sc1.body, ordxctptncode: "13", canccnfqty: "1" } }, TODAY)[0].type, "CANCELED");
  // 하락 부호: change 는 부호 없음 → sign 5 면 음수
  const falling = JSON.parse(TRADE_FRAME);
  falling.body.sign = "5";
  assert.ok(parseLsTrade(falling, TODAY)!.change!.eq(-1050));
});

test("ls 픽스처: stream 섹션(문서 기반)을 기대값대로 파싱한다", () => {
  assert.equal(fx.measured, false);
  assertTrades(fx.frames.map((f) => parseLsTrade(JSON.parse(f), TODAY)!), fx.expected);
  assert.equal(fx.orderBook.measured, false);
  assertBooks(fx.orderBook.frames.map((f) => parseLsOrderBook(JSON.parse(f), TODAY)!), fx.orderBook.expected);
  assert.equal(fx.orderEvents.measured, false);
  assertEvents(fx.orderEvents.frames.flatMap((f) => parseLsOrderEvents(JSON.parse(f), TODAY)), fx.orderEvents.expected);
});

File: go/ls_stream_test.go
package hermetix

// LS증권 스트림 — 가짜 웹소켓 서버로 등록(KOSPI+KOSDAQ 이중)·ACK·체결·호가·통보·해제·재접속 흐름과 픽스처(문서 기반) 파싱을 검증한다.

import (
	"encoding/json"
	"strings"
	"testing"
	"time"
)

type lsMsg struct {
	Header struct {
		Token  string `json:"token"`
		TrType string `json:"tr_type"`
	} `json:"header"`
	Body struct {
		TrCd  string `json:"tr_cd"`
		TrKey string `json:"tr_key"`
	} `json:"body"`
}

func parseLsMsg(t *testing.T, raw string) lsMsg {
	t.Helper()
	var m lsMsg
	if err := json.Unmarshal([]byte(raw), &m); err != nil {
		t.Fatal(err)
	}
	return m
}

func newLsTestStream(t *testing.T) (*LsMarketStream, chan *serverConn) {
	t.Helper()
	srv, connections := newWsServer(t)
	stream := newLsMarketStream(wsURL(srv), func() (string, error) { return "TOKEN", nil })
	t.Cleanup(func() { stream.Close() })
	return stream, connections
}

func takeLs(t *testing.T, conn *serverConn, n int) map[string]lsMsg {
	t.Helper()
	out := map[string]lsMsg{}
	for i := 0; i < n; i++ {
		m := parseLsMsg(t, conn.take(t))
		out[m.Body.TrCd] = m
	}
	return out
}

func TestLsStreamDualSubscribeAndTick(t *testing.T) {
	fx := loadStreamFixture(t, "ls")
	stream, connections := newLsTestStream(t)
	ticks := make(chan TradeTick, 16)
	stream.SubscribeTrades([]string{"KRX:005930"}, func(tick TradeTick) { ticks <- tick })
	stream.Connect()

	conn := waitConn(t, connections, 5*time.Second)
	msgs := takeLs(t, conn, 2)
	for _, tr := range []string{"S3_", "K3_"} {
		m, ok := msgs[tr]
		if !ok || m.Header.Token != "TOKEN" || m.Header.TrType != "3" || m.Body.TrKey != "005930" {
			t.Fatalf("%s subscribe = %+v", tr, m)
		}
	}
	conn.send(t, `{"header":{"tr_cd":"S3_","tr_key":"005930","tr_type":"3","rsp_cd":"00000","rsp_msg":"정상처리되었습니다"}}`)
	conn.send(t, fx.Frames[0])
	tick := waitTick(t, ticks)
	if tick.Symbol != "KRX:005930" || tick.Price.String() != "55550" || tick.Quantity.String() != "1" {
		t.Fatalf("tick = %+v", tick)
	}
	if tick.CumulativeVolume == nil || *tick.CumulativeVolume != 10887 {
		t.Fatalf("cumulative = %v", tick.CumulativeVolume)
	}
	assertDecimalEqual(t, "change", tick.Change, "1050")
	assertDecimalEqual(t, "changeRate", tick.ChangeRate, "0.0193")
	assertDecimalEqual(t, "ask", tick.AskPrice, "55600")
	assertDecimalEqual(t, "bid", tick.BidPrice, "55500")

	// KOSDAQ TR 도 같은 레이아웃으로 전달된다
	conn.send(t, strings.Replace(fx.Frames[0], `"tr_cd":"S3_"`, `"tr_cd":"K3_"`, 1))
	if kosdaq := waitTick(t, ticks); kosdaq.Symbol != "KRX:005930" {
		t.Fatalf("kosdaq tick = %+v", kosdaq)
	}
}

func TestLsStreamOrderEventsBookAndReconnect(t *testing.T) {
	fx := loadStreamFixture(t, "ls")
	stream, connections := newLsTestStream(t)
	books := make(chan OrderBookTick, 4)
	events := make(chan OrderEvent, 8)
	if err := stream.SubscribeOrderBook([]string{"005930"}, func(b OrderBookTick) { books <- b }); err != nil {
		t.Fatal(err)
	}
	if err := stream.SubscribeOrderEvents(func(e OrderEvent) { events <- e }); err != nil {
		t.Fatal(err)
	}
	stream.Connect()
	conn := waitConn(t, connections, 5*time.Second)
	msgs := takeLs(t, conn, 7)
	for _, tr := range []string{"H1_", "HA_"} {
		if m := msgs[tr]; m.Header.TrType != "3" || m.Body.TrKey != "005930" {
			t.Fatalf("%s = %+v", tr, m)
		}
	}
	for _, tr := range lsOrderTRs {
		m, ok := msgs[tr]
		if !ok || m.Header.TrType != "1" || m.Body.TrKey != "" {
			t.Fatalf("%s 등록은 tr_type 1·tr_key 빈 문자열이어야 한다: %+v", tr, m)
		}
	}

	conn.send(t, fx.OrderBook.Frames[0])
	var book OrderBookTick
	select {
	case book = <-books:
	case <-time.After(5 * time.Second):
		t.Fatal("호가가 오지 않음")
	}
	best, _ := book.BestAsk()
	bid, _ := book.BestBid()
	if book.Symbol != "005930" || len(book.Asks) != 10 || best.Price.String() != "72400" || bid.Price.String() != "72300" {
		t.Fatalf("book = %+v", book)
	}

	for _, frame := range fx.OrderEvents.Frames {
		conn.send(t, frame)
	}
	take := func() OrderEvent {
		select {
		case e := <-events:
			return e
		case <-time.After(5 * time.Second):
			t.Fatal("이벤트가 오지 않음")
			return OrderEvent{}
		}
	}
	accepted := take()
	if accepted.Type != OrderAccepted || accepted.OrderID != "86382" || accepted.Symbol != "005930" || *accepted.Side != Buy {
		t.Fatalf("accepted = %+v", accepted)
	}
	assertDecimalEqual(t, "quantity", accepted.Quantity, "2")
	assertDecimalEqual(t, "price", accepted.Price, "60000")
	filled := take()
	if filled.Type != OrderFilled {
		t.Fatalf("filled = %+v", filled)
	}
	assertDecimalEqual(t, "fill qty", filled.Quantity, "1")
	assertDecimalEqual(t, "remaining", filled.RemainingQuantity, "1")
	canceled := take()
	if canceled.Type != OrderCanceled || canceled.OrderID != "88343" || canceled.OriginalOrderID != "88342" || canceled.Symbol != "000020" {
		t.Fatalf("canceled = %+v", canceled)
	}

	// 해제: 호가 tr_type 4, 계좌 tr_type 2
	stream.UnsubscribeOrderBook([]string{"005930"})
	stream.UnsubscribeOrderEvents()
	unsub := takeLs(t, conn, 7)
	if unsub["H1_"].Header.TrType != "4" || unsub["HA_"].Header.TrType != "4" || unsub["SC1"].Header.TrType != "2" {
		t.Fatalf("unsubscribe = %+v", unsub)
	}

	// 재접속: 남은 구독은 없으므로 아무것도 보내지 않아야 한다 — 새 트레이드 구독을 넣고 재접속을 확인
	stream.SubscribeTrades([]string{"000660"}, func(TradeTick) {})
	takeLs(t, conn, 2)
	_ = conn.conn.Close(1000, "bye")
	second := waitConn(t, connections, 10*time.Second)
	again := takeLs(t, second, 2)
	if again["S3_"].Body.TrKey != "000660" || again["K3_"].Body.TrKey != "000660" {
		t.Fatalf("resubscribe = %+v", again)
	}
}

func TestLsParsersEdgeCases(t *testing.T) {
	today := fixtureToday
	control := `{"header":{"tr_cd":"S3_","tr_key":"005930","rsp_cd":"00000","rsp_msg":"ok"}}`
	if _, ok := ParseLsTrade(parseFrame(t, control), today); ok {
		t.Fatal("제어 프레임은 파싱하지 않는다")
	}
	if len(ParseLsOrderEvents(parseFrame(t, control), today)) != 0 {
		t.Fatal("제어 프레임은 이벤트가 아니다")
	}
	falling := `{"header":{"tr_cd":"K3_","tr_key":"122870"},"body":{"chetime":"151055","sign":"5","change":"200","drate":"-0.37","price":"53900","cvolume":"438","volume":"14738346","offerho":"54000","bidho":"53900","shcode":"122870"}}`
	tick, ok := ParseLsTrade(parseFrame(t, falling), today)
	if !ok || tick.Symbol != "122870" {
		t.Fatalf("tick = %+v", tick)
	}
	assertDecimalEqual(t, "change (falling)", tick.Change, "-200")
	assertDecimalEqual(t, "changeRate", tick.ChangeRate, "-0.0037")

	modify := `{"header":{"tr_cd":"SC2"},"body":{"ordxctptncode":"12","ordno":"86383","orgordno":"86382","shtnIsuno":"A005930","bnstp":"2","ordqty":"1","ordprc":"70000","mdfycnfqty":"1","mdfycnfprc":"70000","unercqty":"1","exectime":"100045203"}}`
	events := ParseLsOrderEvents(parseFrame(t, modify), today)
	if len(events) != 1 || events[0].Type != OrderModified || events[0].OriginalOrderID != "86382" || events[0].Price.String() != "70000" {
		t.Fatalf("modified = %+v", events)
	}
	reject := `{"header":{"tr_cd":"SC4"},"body":{"ordno":"90000","orgordno":"0","shtnIsuno":"A005930","bnstp":"2","rjtqty":"1","ordprc":"1","msgcode":"0040","exectime":"100100000"}}`
	events = ParseLsOrderEvents(parseFrame(t, reject), today)
	if len(events) != 1 || events[0].Type != OrderRejected || events[0].Reason != "0040" || events[0].OriginalOrderID != "" {
		t.Fatalf("rejected = %+v", events)
	}
	// ordxctptncode 가 tr_cd 보다 우선한다
	precedence := strings.Replace(modify, `"tr_cd":"SC2"`, `"tr_cd":"SC1"`, 1)
	if events = ParseLsOrderEvents(parseFrame(t, precedence), today); len(events) != 1 || events[0].Type != OrderModified {
		t.Fatalf("precedence = %+v", events)
	}
}

func TestLsStreamFixture(t *testing.T) {
	fx := loadStreamFixture(t, "ls")
	if fx.Measured {
		t.Fatal("ls 픽스처는 문서 기반(measured=false)이어야 한다")
	}
	var ticks []TradeTick
	for _, raw := range fx.Frames {
		if tick, ok := ParseLsTrade(parseFrame(t, raw), fixtureToday); ok {
			ticks = append(ticks, tick)
		}
	}
	assertStreamTicks(t, ticks, fx)

	var books []OrderBookTick
	for _, raw := range fx.OrderBook.Frames {
		if book, ok := ParseLsOrderBook(parseFrame(t, raw), fixtureToday); ok {
			books = append(books, book)
		}
	}
	assertBooks(t, books, fx.OrderBook)

	var events []OrderEvent
	for _, raw := range fx.OrderEvents.Frames {
		events = append(events, ParseLsOrderEvents(parseFrame(t, raw), fixtureToday)...)
	}
	assertOrderEvents(t, events, fx.OrderEvents)
}
